src/catalog/catalog.cpp

```diff
@@ -0,0 +1,556 @@
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "catalog/attr_def_handle.h"
+#include "catalog/catalog.h"
+#include "catalog/class_handle.h"
+#include "catalog/database_handle.h"
+#include "catalog/settings_handle.h"
+#include "catalog/tablespace_handle.h"
+#include "loggers/catalog_logger.h"
+#include "storage/storage_defs.h"
+#include "transaction/transaction_manager.h"
+
+namespace terrier::catalog {
+
+std::shared_ptr<Catalog> terrier_catalog;
+
+Catalog::Catalog(transaction::TransactionManager *txn_manager) : txn_manager_(txn_manager), oid_(START_OID) {
+  CATALOG_LOG_TRACE("Creating catalog ...");
+  Bootstrap();
+  CATALOG_LOG_TRACE("=======Finished Bootstrapping ======");
+}
+
+void Catalog::CreateDatabase(transaction::TransactionContext *txn, const std::string &name) {
+  db_oid_t new_db_oid = db_oid_t(GetNextOid());
+  Catalog::AddEntryToPGDatabase(txn, new_db_oid, name);
+  BootstrapDatabase(txn, new_db_oid);
+}
+
+void Catalog::DeleteDatabase(transaction::TransactionContext *txn, const std::string &db_name) {
+  // get database handle
+  auto db_handle = GetDatabaseHandle();
+  auto db_entry = db_handle.GetDatabaseEntry(txn, db_name);
+  auto oid = db_entry->GetOid();
+  // remove entry from pg_database
+  db_handle.DeleteEntry(txn, db_entry);
+
+  // TODO(pakhtar):
+  // - delete all the tables
+
+  // drop database local tables
+  // - pg_attribute
+  // - pg_namespace
+  // - pg_class
+  // - pg_type
+  // - pg_attrdef
+
+  map_.erase(oid);
+  name_map_.erase(oid);
+}
+
+table_oid_t Catalog::CreateTable(transaction::TransactionContext *txn, db_oid_t db_oid, const std::string &table_name,
+                                 const Schema &schema) {
```

Comment:
Is this method used to create a catalog table or user table, or both? Right now it is only used in the `CreateUserTableTest`. Also, I don't think you can call this method concurrently, because it is modifying `map_` and `name_map_`, which are not concurrent data structures. Is it designed that way?

Reply:
It is just for creation of user tables and currently just used in the test. 

Issue #311 notes the concurrency issue with maps.

Reply:
Maybe rename it to `CreateUserTable` to make it more explicit?

Reply:
Renamed to CreateUserTable